### DISCUSSION TITLE
Automate "Rename tab"

### DISCUSSION THREAD

**Discussion by amariusz** [freeplane/freeplane#1688] [Help] [answered]
I wonder if there's a programatic way to invoke "Rename tab". My use case would be hiding part of the filename with timestamp (long suffix) or generally hiding patterns that match regular expression in order to have short and clean tab names. It's probably not available for scripts via API?

Alternatively I was hoping to achive this result using find&replace on config file when FreePlane is off. It seems however that tab names are stored in `auto.properties` file `openedNow_1.3.04` binary property.

Also, "Rename tab" is not present in Command search or within Menus.

**Comment by euu2021** [comment]
That's an interesting topic. There are still no methods in the public API to interact with the tab name. I tried to use the internal methods, with no success (but, in reality, I still don't understand how to use internal methods, in general).

The only way that I've seen to actually have some interaction with the tab name is this script mentioned by Edo: https://github.com/freeplane/freeplane/issues/865#issuecomment-1648760926
But, the name is not retained, as he explained there. 

What I managed to do is to retrieve the name of the tab (assuming that you have the lib of infonode, as explained by Edo):

```
import javax.swing.SwingUtilities
import net.infonode.docking.title.DockingWindowTitleProvider;
import net.infonode.docking.DockingWindow;


//To get the map view name in the variable mapViewName
def rootWindow = SwingUtilities.getRootPane(ui.frame).contentPane.components[0].components.find{it.class.name == 'net.infonode.docking.RootWindow'}
def tabWindow =  rootWindow
do{
    tabWindow = tabWindow.getLastFocusedChildWindow()
} while (tabWindow && !tabWindow.insideTab())
if(!tabWindow) {
    def msg = 'no window selected'
    c.statusInfo = msg
    return msg
}
def tabWindowTab = tabWindow.getTab()
if(tabWindowTab.showing ) {
    mapViewName = tabWindowTab.getText()
 } else {
    def msg = "selected view "${tabWindow.title}" is not inside a group of tabs"
    c.statusInfo = msg
    return msg
}
```

**Comment by amariusz** [comment, marked as answer]
Thanks @euu2021, that turned out to be really helpful :)

> The only way that I've seen to actually have some interaction with the tab name is this script mentioned by Edo: [#865 (comment)](https://github.com/freeplane/freeplane/issues/865#issuecomment-1648760926)
> But, the name is not retained, as he explained there.

I saw this thread before and skimmed it but couldn't find answer to my issue. This this time I've also looked at linked pull request and found this very useful comment by @EdoFro 

https://github.com/freeplane/freeplane/pull/1353#issuecomment-1688293804

> NOTE TO MYSELF: just for the record (and to look at that Freeplane code):
> 
> this script changes a mind map's tab's title:
> 
> ```groovy
> node.mindMap.setName("My custom Name")
> ```
> 
> It doesn't show the change in the title right after runing it, but if you change the focused mapview, then it will change.

![240213_082558_setName](https://github.com/freeplane/freeplane/assets/20703230/1cb86fe8-a488-49b8-a039-ca6a8fc8a873)

So indeed FreePlane API offers a method to change tab name!
As @EdoFro mentioned it has some glitches and intricacies but it works most of the time.

My observations with Freeplane 1.11.10:

- After invoking it it's necessary to switch to another tab to reflect the change.
- If you rename tab name manually (right click, rename) this method won't affect the tab name. It will still affect the window title though. To undo manual name to allow setName to work, manually set name to empty (delete contents of Input dialog).
- Sometimes I get error "Error executing the script: map org.freeplane.features.map.mindmapmode.MMapModel@7de846ff does not seem to be opened" when invoking setName. It's gone after reloading mindmap.
- renames with setName are persistent across FreePlane restarts.

This solves my issue. Thanks again!

**Reply by amariusz** [reply]
... and BTW the other solution that came to my mind before was using symbollic link with shorter name for files I commonly use. This approach also seems to work well (on Linux at least).

**Reply by euu2021** [reply]
Nice. I forgot about that setName method.
